### ISSUE TITLE
LIBERO eval: finetuned OpenVLA checkpoint approaches the object but fails to grasp (0/2 episodes)

### ISSUE THREAD

**Issue by SoyGema** [lovellai-dev/odyssey#61]
Tracking a validation finding from **PR #46** (`feat/libero-eval`, LIBERO eval integration), observed running the single-agent example on a GCP L4.

https://github.com/user-attachments/assets/96304c58-511e-440d-bf2e-b51b3791ee34

## Status: eval pipeline validated end-to-end ✅
The hard part works. On `examples/franka-libero/mission.yaml` (suite `libero_object`, `task_id 0`):
- Dependency install resolved (LIBERO co-installs with the OpenVLA stack in a dedicated `env_pilot_libero`; `transformers` stays at 4.40.x, `robosuite` 1.4.0).
- `LiberoRunner` runs end-to-end: model loads, env builds, episodes run and are scored, videos are captured.

## Symptom
The published, task-specific checkpoint `openvla/openvla-7b-finetuned-libero-object` scores **0/2** episodes (`return=0.000`, `FAIL`). In the rollout videos, **the arm moves toward the object but fails to grasp it cleanly** — in both episodes. (Videos to be attached below.)

## Hypothesis
A checkpoint fine-tuned **on this exact suite** is reported at ~70–90% success, so 0/2 is suspicious. However, a code review shows the integration is **faithful to OpenVLA's reference `run_libero_eval.py` on every axis we'd expect to cause this**:

| Axis | `libero.py` | OpenVLA reference | Match |
|---|---|---|---|
| `predict_action` gripper range `[0,1]` | raw output | raw output | ✅ |
| gripper: normalize → binarize → invert | `_libero_action` | `normalize_gripper_action` + `invert_gripper_action` | ✅ |
| image 180° flip | `[::-1, ::-1]` | same | ✅ |
| camera resolution | 256×256 | 256×256 | ✅ |
| settle before acting | 10 no-ops, gripper open | `num_steps_wait` | ✅ |
| `unnorm_key` | `libero_object` | per-suite | ✅ |

So the two leading explanations are:

1. **Statistical variance** — 2 episodes is too small a sample; a ~80% checkpoint failing the grasp (the hard sub-task) twice in a row is plausible. **Action:** re-run with `num_episodes: 10` for a real `success_rate`.
2. **Center-crop preprocessing (leading suspect if it's systematic)** — the `openvla-7b-finetuned-libero-*` checkpoints were trained with `center_crop=True`. If `make_openvla_policy` does not replicate that crop, the image is slightly off-center → the policy approaches but grasps imprecisely, which matches the observed "reaches but misses" behavior. **Action:** verify the crop path in `make_openvla_policy` against the reference.

## Next steps
- [ ] Re-run `libero_object` / `task_id 0` with `num_episodes: 10` and report `success_rate`.
- [ ] If success is ~0/10, audit `center_crop` in `make_openvla_policy` vs `run_libero_eval.py`.
- [ ] Attach rollout videos (below).

## Environment
GCP L4 (24 GB) · `env_pilot_libero` · `MUJOCO_GL=osmesa` (EGL failed headless) · `robosuite==1.4.0` · `transformers` 4.40.x · LIBERO installed via clone + `.pth` (namespace package).

Refs: PR #46.

**Comment by SoyGema** [comment]
### Update: confirmed systematic (0/10) — root cause = missing eval-time center crop → fixed

Re-ran `libero_object` / `task_id 0` with `num_episodes: 10`: **0/10, all `return=0.000`**. That rules out variance — it's systematic.

**Root cause (confirmed in code):** `make_openvla_policy` (and `VLARuntime`) fed the full frame straight to the processor, with **no center crop**. The `openvla-7b-finetuned-libero-*` checkpoints were fine-tuned with `image_aug=True` (RandomResizedCrop ~0.9), so OpenVLA's `run_libero_eval.py` applies a matching deterministic **center crop (crop_scale 0.9)** at inference. Without it the input FOV is wider than training → the policy's spatial mapping is off → **arm approaches but misses the grasp** (exactly the observed behavior).

**Fix (`5f7085d` on `feat/libero-eval`):** added `_center_crop_image` and applied it in both policy paths, gated by config `center_crop` (default `True`, the OpenVLA convention) + `crop_scale` (default `0.9`). ruff + mypy + unit tests green.

**Next:** re-run the 10-episode eval on the L4 with the fix and confirm `success_rate` jumps to the expected ~70–90% range. Will update here with the number.

**Comment by SoyGema** [comment]
### Update: center-crop fix applied but does NOT resolve it — next step is a reference bisection

Pulled the center-crop fix (`5f7085d`) onto the L4 and re-ran the single-agent eval (`libero_object`, `task_id 0`) **with the crop active**. Still failing: **0 successes in the first 3 episodes** (run in progress; full 10-episode number pending). So the eval-time center crop, while correct to add, is **not sufficient** on its own — there's at least one more factor.

**What we've actually verified so far**
- Pipeline runs end-to-end (deps, model load, env, scoring, video). ✅
- Code path matches OpenVLA's `run_libero_eval.py` on: 180° image flip, gripper normalize+binarize+invert, camera 256², 10 settle no-ops, `unnorm_key`, raw `predict_action` gripper `[0,1]`. ✅
- Center crop (crop_scale 0.9) now applied in both policy paths (`5f7085d`). ✅
- Observed behaviour (from rollout videos): the arm **approaches the object but misses the grasp** — consistent across episodes, both before and after the crop fix.

**Hypotheses NOT yet tested**
1. **Render backend (leading, untested):** we run with `MUJOCO_GL=osmesa` (CPU software render) because EGL failed headless on this VM. osmesa may produce images subtly out-of-distribution vs the GPU-rendered frames the checkpoint trained on → policy approaches but lacks fine precision. **Not verified.**
2. **Residual preprocessing discrepancy (untested):** e.g. PIL `BILINEAR` crop-resize vs OpenVLA's `tf.image.crop_and_resize` interpolation, RGB channel order, or resize path.

**Planned next step — bisect "our runner vs the environment"**
Run OpenVLA's own `experiments/robot/libero/run_libero_eval.py` on the **same checkpoint** in the **same env (osmesa, robosuite 1.4.0)**:
- if it **also scores ~0** → environmental (render backend / robosuite version), not our runner → fix by recovering EGL;
- if it **scores 70–90%** → a residual gap in our preprocessing → compare frame-by-frame.

Will post the reference-eval result next.

**Comment by SoyGema** [comment]
### Paused here — next (untested) steps for whoever picks this up

Current state: the eval pipeline works end-to-end, but `libero_object` scores ~0 with a task-specific checkpoint. The center-crop fix (`5f7085d`) is in and did **not** resolve it. Everything below is **not yet tested**:

1. **Bisection — run OpenVLA's own eval as ground truth.** Run `~/openvla/experiments/robot/libero/run_libero_eval.py` on the **same checkpoint** (`openvla/openvla-7b-finetuned-libero-object`) in the **same env** (osmesa, robosuite 1.4.0):
   - also ~0 → **environmental** (render backend / robosuite version), not our `LiberoRunner`;
   - 70–90% → a **residual preprocessing gap** in our runner.
2. **If environmental → recover EGL.** We're on `MUJOCO_GL=osmesa` (CPU software render) only because EGL failed headless on this VM. osmesa *may* produce out-of-distribution frames vs the GPU-rendered training data. Getting `MUJOCO_GL=egl` working would test this and also speed the eval up a lot. **The osmesa-fidelity hypothesis is unverified.**
3. **If our runner → compare frame-by-frame.** Prime suspects: crop interpolation (PIL `BILINEAR` vs OpenVLA's `tf.image.crop_and_resize`), RGB channel order, and the resize path into the processor.

Note: the multi-agent mission (`mission-multiagent.yaml`) shares the same OpenVLA policy path (`VLARuntime`), so it's expected to show the same `success_rate` behavior until this is resolved — a separate multi-agent-plumbing validation is tracked on PR #46.

No further testing today; picking up from step 1 next session.

**Comment by SoyGema** [comment]
### Episode horizon doubled (600 → 1200) based on rollout observation

Watching the single-agent rollout videos, the arm moves **smoothly toward the target but runs out of steps** — it reaches the object and doesn't finish the grasp before the episode ends, rather than flailing/oscillating. That points at horizon (and/or per-step action scale) rather than pure visual imprecision.

Raised `LiberoRunner` default `max_steps_per_episode` **600 → 1200** (`0582038` on `feat/libero-eval`; still overridable via `config.max_steps_per_episode`). For reference, OpenVLA's `run_libero_eval.py` uses ~280 steps for `libero_object`, so if the arm needs >600 it's already moving slower than the reference expects — which also keeps **action-scale / un-normalization** on the suspect list.

**Next re-run:** `libero_object` with the 1200 horizon. Interpretation:
- success rate improves → horizon/speed was the limiter (document a higher default);
- still ~0 but the arm now visibly **closes on the object** → getting closer; look at action scale;
- still ~0 and no progress → back to the render/preprocessing bisection (untested OpenVLA-reference run above).

(Same commit also fixed the unrelated CI mypy failure: `Image.BILINEAR` → `Image.Resampling.BILINEAR`.)

**Comment by SoyGema** [comment]
## Resolved: the arm now grasps the target object

**Update — the "approaches but fails to grasp" symptom is gone.** On the latest eval runs the policy reaches and **grasps** the target object cleanly (see attached rollout video). The original 0/2 "fails to grasp" behaviour this issue was opened for no longer reproduces.

### Key correction: the render backend was NOT the cause
The grasp works in **both** `osmesa` (CPU render) and `EGL` (GPU render). So the earlier hypothesis — that osmesa's render caused the grasp failure via visual distribution shift — does **not** explain it. Switching backend is not what fixed the grasp. (EGL is still worth keeping for GPU-side render **speed**, just not as the grasp fix.)

Setup done along the way on the eval VM: recovered the NVIDIA kernel driver (kernel/driver module mismatch after a kernel upgrade) and installed the missing NVIDIA GL/EGL userspace runtime (`libnvidia-gl-580-server`), which is what made hardware EGL available at all.

### What remains (tracked under PR #46, not this issue)
The episode can still be marked FAIL because LIBERO success is the full BDDL goal — *"pick up X **and place it in the basket**"* — not the grasp alone. That residual place-completion question (horizon vs policy) is being investigated separately: re-running with an explicit `max_steps_per_episode: 2000` to rule out simply running out of steps mid-place, and confirming the grasped object matches the task instruction.

Closing this one as the grasp failure is resolved; video attached. Place-completion follow-up lives with the eval integration in #46.

**Comment by SoyGema** [comment]
<img width="770" height="757" alt="Image" src="https://github.com/user-attachments/assets/e0834332-1f3a-4069-a82e-781b2ffce107" />

**Comment by SoyGema** [comment]
Quite an exploration : learnt about how VLA models are fragile to visualization pixels
